integration/querydsl/blaze-persistence-querydsl-expressions/src/main/java/com/blazebit/persistence/querydsl/WindowFirstLast.java

```diff
@@ -0,0 +1,146 @@
+
+    private static final long serialVersionUID = 4107262569593794721L;
+
+    private static final String ORDER_BY = "order by ";
+
+    private final List<OrderSpecifier<?>> orderBy = new ArrayList<OrderSpecifier<?>>();
+
+    @Nullable
+    private transient volatile SimpleExpression<T> value;
+
+    private final Expression<T> target;
+
+    private final boolean first;
+
+    public WindowFirstLast(WindowOver<T> target, boolean first) {
+        super(target.getType());
+        this.target = target;
+        this.first = first;
+    }
+
+    @Override
+    public <R,C> R accept(Visitor<R,C> v, C context) {
+        return getValue().accept(v, context);
+    }
+
+    /**
+     * Add an order by clause to this window definition.
+     *
+     * @param orderBy order by expression
+     * @return this window definition
+     * @since 1.5.0
+     */
+    public WindowFirstLast<T> orderBy(ComparableExpressionBase<?> orderBy) {
+        value = null;
+        this.orderBy.add(orderBy.asc());
+        return this;
+    }
+
+    /**
+     * Add an order by clause to this window definition.
+     *
+     * @param orderBy order by expressions
+     * @return this window definition
+     * @since 1.5.0
+     */
+    public WindowFirstLast<T> orderBy(ComparableExpressionBase<?>... orderBy) {
+        value = null;
+        for (ComparableExpressionBase<?> e : orderBy) {
+            this.orderBy.add(e.asc());
+        }
+        return this;
+    }
+
+    /**
+     * Add an order by clause to this window definition.
+     *
+     * @param orderBy order by expression
+     * @return this window definition
+     * @since 1.5.0
+     */
+    public WindowFirstLast<T> orderBy(OrderSpecifier<?> orderBy) {
+        value = null;
+        this.orderBy.add(orderBy);
+        return this;
+    }
+
+    /**
+     * Add an order by clause to this window definition.
+     *
+     * @param orderBy order by expressions
+     * @return this window definition
+     * @since 1.5.0
+     */
+    public WindowFirstLast<T> orderBy(OrderSpecifier<?>... orderBy) {
+        value = null;
+        Collections.addAll(this.orderBy, orderBy);
+        return this;
+    }
+
+    SimpleExpression<T> getValue() {
+        if (value == null) {
+            if (orderBy.isEmpty()) {
+                // TODO this check should be static
+                throw new IllegalStateException("No order by arguments given");
+            }
+            ImmutableList.Builder<Expression<?>> args = ImmutableList.builder();
+            StringBuilder builder = new StringBuilder();
+            builder.append("{0} keep (dense_rank ");
+            args.add(target);
+            builder.append(first ? "first " : "last ");
+            builder.append(ORDER_BY);
+            builder.append("{1}");
+            args.add(ExpressionUtils.orderBy(orderBy));
+            builder.append(")");
+            value = Expressions.template(target.getType(), builder.toString(), (List<Expression<?>>) args.build());
```

Comment:
Please create two static final string templates for the first and last variant and conditionally use that here instead of building that up on demand.

Reply:
```suggestion
            this.value = value = Expressions.template(target.getType(), builder.toString(), (List<Expression<?>>) args.build());
```